src/cart/src/gurt/hash.c

```diff
@@ -690,58 +788,105 @@ d_hash_table_create_inplace(uint32_t feats, unsigned int bits, void *priv,
 	htable->ht_ops	 = hops;
 	htable->ht_priv	 = priv;
 
-	D_ALLOC_ARRAY(buckets, nr);
-	if (buckets == NULL)
-		return -DER_NOMEM;
+	if (hops->hop_rec_hash == NULL && !(feats & D_HASH_FT_NOLOCK)) {
+		htable->ht_feats |= D_HASH_FT_GLOCK;
+		D_WARN("The d_hash_table_ops_t->hop_rec_hash() callback is "
+			"not provided!\nTherefore the whole hash table locking "
+			"will be used for backward compatibility.\n");
+	}
+
+	D_ALLOC_ARRAY(htable->ht_buckets, nr);
+	if (htable->ht_buckets == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
 
 	for (i = 0; i < nr; i++)
-		D_INIT_LIST_HEAD(&buckets[i].hb_head);
+		D_INIT_LIST_HEAD(&htable->ht_buckets[i].hb_head);
 
-	htable->ht_buckets = buckets;
-	rc = ch_lock_init(htable);
-	if (rc != 0) {
-		D_FREE(buckets);
-		return rc;
+	if (htable->ht_feats & D_HASH_FT_NOLOCK)
+		D_GOTO(out, rc = 0);
+
+	if (htable->ht_feats & D_HASH_FT_GLOCK) {
+		if (htable->ht_feats & D_HASH_FT_MUTEX)
+			rc = D_MUTEX_INIT(&htable->ht_lock.mutex, NULL);
+		else if (htable->ht_feats & D_HASH_FT_RWLOCK)
+			rc = D_RWLOCK_INIT(&htable->ht_lock.rwlock, NULL);
+		else
+			rc = D_SPIN_INIT(&htable->ht_lock.spin,
+					 PTHREAD_PROCESS_PRIVATE);
+		if (rc)
+			D_GOTO(free_buckets, rc);
+	} else {
+		D_ALLOC_ARRAY(htable->ht_locks, nr);
+		if (htable->ht_locks == NULL)
+			D_GOTO(free_buckets, rc = -DER_NOMEM);
+
+		for (i = 0; i < nr; i++) {
+			if (htable->ht_feats & D_HASH_FT_MUTEX)
+				rc = D_MUTEX_INIT(&htable->ht_locks[i].mutex,
+						  NULL);
+			else if (htable->ht_feats & D_HASH_FT_RWLOCK)
+				rc = D_RWLOCK_INIT(&htable->ht_locks[i].rwlock,
+						   NULL);
```

Comment:
What would happen if D_HASH_FT_RWLOCK was used in combination with the LRU?  Would that be something that could be checked here?

Reply:
It can be used. What the problem with this you see?
The bad things here will be just in most places a write lock will be used. Only addref/decref will use read lock.

Reply:
So find has changed from read lock to write lock?  That was my concern.

Reply:
Yes, It check now. If LRU bit is set it use write lock.